File: thud/experiments/az_layout_check.py
# Copyright 2026 The Thud-on-OpenSpiel authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Does OpenSpiel's Python AlphaZero model learn Thud better with the planes as channels?

OpenSpiel's flax model reshapes each observation to game.observation_tensor_shape(), planes
first, and hands it to flax's nn.Conv, which takes the LAST axis as channels
(model_linen.py:209/217). For Thud's (6, 15, 15) the convolution then slides over (plane,
row), with the 15 board columns as channels. This trains OpenSpiel's own model and update
step, unmodified, in two variants on the same data:

  as_is  the observation as AlphaZero feeds it today, input shape (6, 15, 15);
  fixed  the same observation transposed to (15, 15, 6), input shape (15, 15, 6), so the
         unchanged reshape yields planes-last and the convolution runs over the board.

The task is purely about the board, and one AlphaZero needs: the value head learns whether
the side to move has a capture (+1) or not (-1); the policy target is uniform over the
capturing moves if there are any, else over all legal moves. Positions come from random
games, train and test from different games.

  python3 az_layout_check.py --variant as_is --seed 1
  python3 az_layout_check.py --variant fixed --seed 1

Prints one JSON line per evaluation. Needs pyspiel and OpenSpiel's JAX dependencies.

  python3 az_layout_check.py --export DIR

writes what az_layout_check.cc, the C++ counterpart, needs to train OpenSpiel's C++
model on exactly these positions and batches (seeds 1-3); see that file.
"""

import argparse
import json
import os
import pickle
import random
import time

import jax
import jax.numpy as jnp
import numpy as np
import pyspiel

from open_spiel.python.algorithms.alpha_zero import model_linen
from open_spiel.python.algorithms.alpha_zero import utils

SHAPE = (6, 15, 15)  # ThudGame::ObservationTensorShape()
TROLLS_TO_MOVE_PLANE = 3


def positions(game, seeds, every, games=None):
  """Every `every`-th position of random games: (observation, legal, capturing).

  If `games` is a list, each game's actions are appended to it.
  """
  data = []
  for seed in seeds:
    rng, state, turn, actions = random.Random(seed), game.new_initial_state(), 0, []
    while not state.is_terminal():
      legal = state.legal_actions()
      if turn % every == 0:
        captures = [a for a in legal if state.action_to_string(a).endswith("x")]
        data.append((np.asarray(state.observation_tensor(), np.float32),
                     np.asarray(legal), np.asarray(captures, dtype=np.int64)))
      actions.append(rng.choice(legal))
      state.apply_action(actions[-1])
      turn += 1
    if games is not None:
      games.append(actions)
  return data


def load_data(game, cache):
  if os.path.exists(cache):
    with open(cache, "rb") as f:
      return pickle.load(f)
  # Every 7th turn: an odd step, so both sides to move are sampled.
  data = positions(game, range(400), 7), positions(game, range(10_000, 10_100), 7)
  with open(cache, "wb") as f:
    pickle.dump(data, f)
  return data


def checksums(data):
  """Counts and sums that any exact copy of the positions must reproduce."""
  # Two planes hold fractions, so the planes are summed in float64 and rounded.
  obs = np.stack([o for o, _, _ in data]).reshape(-1, *SHAPE).astype(np.float64)
  return {"positions": len(data),
          "legal": sum(len(l) for _, l, _ in data),
          "legal_sum": int(sum(l.sum() for _, l, _ in data)),
          "captures": sum(len(c) for _, _, c in data),
          "captures_sum": int(sum(c.sum() for _, _, c in data)),
          "plane_sums": [round(float(s), 3) for s in obs.sum(axis=(0, 2, 3))]}


def export(game, directory, steps, batch_size, seeds):
  """Writes what the C++ counterpart needs to train on exactly these positions.

  train_games.txt and test_games.txt hold one game's actions per line, to replay;
  summary.json the positions' checksums; batches_seed<S>.bin the batch indices seed S
  draws, steps x batch_size little-endian int32, in order.
  """
  os.makedirs(directory, exist_ok=True)
  summary = {"every": 7, "steps": steps, "batch_size": batch_size}
  for name, seeds_of_games in (("train", range(400)), ("test", range(10_000, 10_100))):
    games = []
    summary[name] = checksums(positions(game, seeds_of_games, 7, games))
    with open(os.path.join(directory, f"{name}_games.txt"), "w") as f:
      f.writelines(" ".join(map(str, actions)) + "\n" for actions in games)
  for seed in seeds:  # As main() draws them.
    rng = np.random.default_rng(seed)
    indices = np.stack([rng.integers(summary["train"]["positions"], size=batch_size)
                        for _ in range(steps)])
    indices.astype("<i4").tofile(os.path.join(directory, f"batches_seed{seed}.bin"))
  with open(os.path.join(directory, "summary.json"), "w") as f:
    json.dump(summary, f, indent=1)
  print(json.dumps(summary))


def model_input(observation, variant):
  if variant == "fixed":
    return observation.reshape(SHAPE).transpose(1, 2, 0).ravel()
  return observation


def batch(data, indices, variant, num_actions):
  n = len(indices)
  mask = np.zeros((n, num_actions), bool)
  policy = np.zeros((n, num_actions), np.float32)
  value = np.empty(n, np.float32)
  for row, i in enumerate(indices):
    _, legal, captures = data[i]
    target = captures if len(captures) else legal
    mask[row, legal] = True
    policy[row, target] = 1 / len(target)
    value[row] = 1.0 if len(captures) else -1.0
  obs = np.stack([model_input(data[i][0], variant) for i in indices])
  return utils.TrainInput(observation=jnp.asarray(obs), legals_mask=jnp.asarray(mask),
                          policy=jnp.asarray(policy), value=jnp.asarray(value))


def evaluate(model, predict, test, variant, num_actions):
  """Value sign accuracy, and policy mass on capturing moves, per side to move."""
  stats = {}
  for start in range(0, len(test), 256):
    idx = list(range(start, min(start + 256, len(test))))
    b = batch(test, idx, variant, num_actions)
    logits, value = predict(model._state.params, model._state.batch_stats,  # pylint: disable=protected-access
                            b.observation)
    logits = jnp.where(b.legals_mask, logits, jnp.finfo(jnp.float32).min)
    probs = np.asarray(jax.nn.softmax(logits, axis=-1))
    for row, i in enumerate(idx):
      obs, _, captures = test[i]
      side = "trolls" if obs.reshape(SHAPE)[TROLLS_TO_MOVE_PLANE].any() else "dwarfs"
      s = stats.setdefault(side, {"n": 0, "right": 0, "with_capture": 0, "mass": 0.0})
      s["n"] += 1
      s["right"] += (float(value[row]) > 0) == bool(len(captures))
      if len(captures):
        s["with_capture"] += 1
        s["mass"] += float(probs[row, captures].sum())
  return {side: {"value_accuracy": round(s["right"] / s["n"], 4),
                 "capture_mass": round(s["mass"] / max(1, s["with_capture"]), 4)}
          for side, s in sorted(stats.items())}


def main():
  parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
  parser.add_argument("--variant", choices=["as_is", "fixed"])
  parser.add_argument("--seed", type=int, default=1)
  parser.add_argument("--steps", type=int, default=1500)
  parser.add_argument("--eval_every", type=int, default=250)
  parser.add_argument("--batch_size", type=int, default=128)
  parser.add_argument("--nn_width", type=int, default=32)
  parser.add_argument("--nn_depth", type=int, default=2)
  parser.add_argument("--cache", default="/tmp/az_layout_check_positions.pkl")
  parser.add_argument("--export", metavar="DIR",
                      help="instead of training, write the data for az_layout_check.cc")
  args = parser.parse_args()

  game = pyspiel.load_game("thud")
  if args.export:
    export(game, args.export, args.steps, args.batch_size, seeds=(1, 2, 3))
    return
  if not args.variant:
    parser.error("--variant is required unless --export is given")
  train, test = load_data(game, args.cache)
  num_actions = game.num_distinct_actions()
  shape = SHAPE if args.variant == "as_is" else (15, 15, 6)
  model = model_linen.Model.build_model(
      "resnet", shape, num_actions, args.nn_width, args.nn_depth, weight_decay=1e-4,
      learning_rate=1e-3, path="/tmp", seed=args.seed)
  kernel = model._state.params["ConvBlock_0"]["Conv_0"]["kernel"].shape  # pylint: disable=protected-access
  apply_fn = model._state.apply_fn  # pylint: disable=protected-access

  @jax.jit
  def predict(params, batch_stats, observations):
    one = lambda o: apply_fn({"params": params, "batch_stats": batch_stats}, o,
                             training=False, mutable=False)
    return jax.vmap(one, axis_name="batch")(observations)

  baselines = {}  # A constant value guess, and a uniform policy, per side to move.
  for side, trolls in (("dwarfs", False), ("trolls", True)):
    mine = [(l, c) for o, l, c in test
            if o.reshape(SHAPE)[TROLLS_TO_MOVE_PLANE].any() == trolls]
    open_ = [len(c) / len(l) for l, c in mine if len(c)]
    share = len(open_) / len(mine)
    baselines[side] = {"positions": len(mine), "value_accuracy": round(max(share, 1 - share), 4),
                       "capture_mass": round(float(np.mean(open_)), 4)}
  head = {"variant": args.variant, "seed": args.seed, "first_kernel": kernel,
          "train": len(train), "test": len(test), "baselines": baselines}
  print(json.dumps(head), flush=True)
  rng, start = np.random.default_rng(args.seed), time.time()
  for step in range(1, args.steps + 1):
    losses = model.update(batch(train, rng.integers(len(train), size=args.batch_size),
                                args.variant, num_actions))
    if step % args.eval_every == 0:
      print(json.dumps({"variant": args.variant, "seed": args.seed, "step": step,
                        "seconds": round(time.time() - start),
                        "loss_policy": round(float(losses.policy), 4),
                        "loss_value": round(float(losses.value), 4),
                        **evaluate(model, predict, test, args.variant, num_actions)}),
            flush=True)


if __name__ == "__main__":
  main()

File: thud/experiments/az_layout_check.cc
// Copyright 2026 The Thud-on-OpenSpiel authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// The C++ counterpart of az_layout_check.py: does OpenSpiel's C++ AlphaZero model
// (LibTorch, planes first) learn Thud's board as well as the Python model does once its
// input is transposed to planes last, and better than the Python model as it is?
//
// It trains upstream's VPNetModel, unmodified, on the Python runs' supervised task, with
// their data and settings: az_layout_check.py --export writes the random games, the
// checksums of the positions taken from them, and each seed's batches. This program
// replays the games with Thud's C++ code, stops unless its positions reproduce the
// checksums, and then trains on the same batches in the same order: resnet 32 x 2,
// learning rate 1e-3, weight decay 1e-4, evaluated on the same per-side measures.
//
//   thud/experiments/build_az_program.sh az_layout_check   # next to libopen_spiel.so
//   python3 thud/experiments/az_layout_check.py --export DATA
//   OMP_NUM_THREADS=3 build-shared/az_layout_check DATA SEED
//
// Prints one JSON line per evaluation, as the Python script does.

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <map>
#include <memory>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#include <torch/torch.h>

#include "open_spiel/abseil-cpp/absl/strings/match.h"
#include "open_spiel/abseil-cpp/absl/strings/str_cat.h"
#include "open_spiel/abseil-cpp/absl/strings/str_format.h"
#include "open_spiel/algorithms/alpha_zero_torch/vpnet.h"
#include "open_spiel/spiel.h"
#include "open_spiel/spiel_utils.h"
#include "open_spiel/utils/json.h"

namespace {

using open_spiel::Action;
using open_spiel::algorithms::torch_az::VPNetModel;

constexpr int kEvery = 7;  // As az_layout_check.py samples.
constexpr int kPlanes = 6;
constexpr int kTrollsToMovePlane = 3;
constexpr int kNnWidth = 32;
constexpr int kNnDepth = 2;
constexpr double kLearningRate = 1e-3;
constexpr double kWeightDecay = 1e-4;
constexpr int kEvalEvery = 250;
constexpr int kEvalBatch = 256;

struct Position {
  std::vector<float> observation;
  std::vector<Action> legal;
  std::vector<Action> captures;
  bool trolls_to_move;
};

// Every kEvery-th position of each game, replayed from its actions.
std::vector<Position> Replay(const open_spiel::Game& game, const std::string& path) {
  std::ifstream file(path);
  if (!file) open_spiel::SpielFatalError(absl::StrCat("Cannot read ", path));
  std::vector<Position> positions;
  std::string line;
  while (std::getline(file, line)) {
    std::istringstream actions(line);
    std::unique_ptr<open_spiel::State> state = game.NewInitialState();
    Action action;
    for (int turn = 0; actions >> action; ++turn) {
      SPIEL_CHECK_FALSE(state->IsTerminal());
      if (turn % kEvery == 0) {
        Position p{state->ObservationTensor(), state->LegalActions(), {}, false};
        for (Action a : p.legal) {
          if (absl::EndsWith(state->ActionToString(a), "x")) p.captures.push_back(a);
        }
        p.trolls_to_move =
            p.observation[kTrollsToMovePlane * p.observation.size() / kPlanes] > 0;
        positions.push_back(std::move(p));
      }
      state->ApplyAction(action);
    }
    SPIEL_CHECK_TRUE(state->IsTerminal());
  }
  return positions;
}

// Stops unless the positions reproduce checksums() of az_layout_check.py, as exported:
// the counts and sums of actions exactly, the plane sums (two planes hold fractions,
// summed in another order) to within 0.01.
void CheckAgainstExport(const std::vector<Position>& positions,
                        const open_spiel::json::Object& expected,
                        const std::string& name) {
  int64_t legal = 0, legal_sum = 0, captures = 0, captures_sum = 0;
  std::vector<double> plane_sums(kPlanes, 0);
  for (const Position& p : positions) {
    legal += p.legal.size();
    for (Action a : p.legal) legal_sum += a;
    captures += p.captures.size();
    for (Action a : p.captures) captures_sum += a;
    const int plane_size = p.observation.size() / kPlanes;
    for (int i = 0; i < p.observation.size(); ++i) {
      plane_sums[i / plane_size] += p.observation[i];
    }
  }
  const std::vector<std::pair<std::string, int64_t>> counts = {
      {"positions", static_cast<int64_t>(positions.size())}, {"legal", legal},       {"legal_sum", legal_sum},
      {"captures", captures},          {"captures_sum", captures_sum}};
  for (const auto& [key, mine] : counts) {
    if (expected.at(key).GetInt() != mine) {
      open_spiel::SpielFatalError(absl::StrCat("The ", name, " positions differ from the "
                                               "export: ", key, " ", mine, " instead of ",
                                               expected.at(key).GetInt()));
    }
  }
  const open_spiel::json::Array& sums = expected.at("plane_sums").GetArray();
  SPIEL_CHECK_EQ(sums.size(), kPlanes);
  for (int plane = 0; plane < kPlanes; ++plane) {
    const double theirs =
        sums[plane].IsInt() ? sums[plane].GetInt() : sums[plane].GetDouble();
    if (std::abs(plane_sums[plane] - theirs) > 0.01) {
      open_spiel::SpielFatalError(absl::StrFormat(
          "The %s positions differ from the export: plane %d sums to %.3f instead of %.3f",
          name, plane, plane_sums[plane], theirs));
    }
  }
}

std::string ReadFile(const std::string& path) {
  std::ifstream file(path);
  if (!file) open_spiel::SpielFatalError(absl::StrCat("Cannot read ", path));
  std::stringstream contents;
  contents << file.rdbuf();
  return contents.str();
}

// The value target: does the side to move have a capture? The policy target: uniform
// over the captures if there are any, else over all legal moves.
VPNetModel::TrainInputs Target(const Position& p) {
  const std::vector<Action>& target = p.captures.empty() ? p.legal : p.captures;
  open_spiel::ActionsAndProbs policy;
  for (Action a : target) policy.push_back({a, 1.0 / target.size()});
  return {p.legal, p.observation, policy, p.captures.empty() ? -1.0 : 1.0};
}

// Value sign accuracy and policy mass on the captures, per side to move, as JSON.
std::string Evaluate(VPNetModel& model, const std::vector<Position>& test) {
  struct Stats {
    int n = 0, right = 0, with_capture = 0;
    double mass = 0;
  };
  std::map<std::string, Stats> stats;
  for (int start = 0; start < test.size(); start += kEvalBatch) {
    std::vector<VPNetModel::InferenceInputs> inputs;
    for (int i = start; i < std::min<int>(start + kEvalBatch, test.size()); ++i) {
      inputs.push_back({test[i].legal, test[i].observation});
    }
    std::vector<VPNetModel::InferenceOutputs> outputs = model.Inference(inputs);
    for (int row = 0; row < outputs.size(); ++row) {
      const Position& p = test[start + row];
      Stats& s = stats[p.trolls_to_move ? "trolls" : "dwarfs"];
      s.n += 1;
      s.right += (outputs[row].value > 0) == !p.captures.empty();
      if (!p.captures.empty()) {
        s.with_capture += 1;
        for (const auto& [action, probability] : outputs[row].policy) {
          for (Action c : p.captures) s.mass += c == action ? probability : 0;
        }
      }
    }
  }
  std::string out;
  for (const auto& [side, s] : stats) {  // std::map: dwarfs, then trolls.
    absl::StrAppend(&out, out.empty() ? "" : ", ",
                    absl::StrFormat("\"%s\": {\"value_accuracy\": %.4f, "
                                    "\"capture_mass\": %.4f}",
                                    side, static_cast<double>(s.right) / s.n,
                                    s.mass / std::max(1, s.with_capture)));
  }
  return out;
}

}  // namespace

int main(int argc, char** argv) {
  if (argc != 3) {
    std::cerr << "Usage: " << argv[0] << " DATA_DIR SEED" << std::endl;
    return 1;
  }
  const std::string data = argv[1];
  const int seed = std::stoi(argv[2]);
  std::shared_ptr<const open_spiel::Game> game = open_spiel::LoadGame("thud");

  const std::vector<Position> train = Replay(*game, data + "/train_games.txt");
  const std::vector<Position> test = Replay(*game, data + "/test_games.txt");
  const open_spiel::json::Object summary =
      open_spiel::json::FromString(ReadFile(data + "/summary.json"))->GetObject();
  CheckAgainstExport(train, summary.at("train").GetObject(), "training");
  CheckAgainstExport(test, summary.at("test").GetObject(), "test");
  const int batch_size = summary.at("batch_size").GetInt();

  // The batches az_layout_check.py draws for this seed, in order.
  std::ifstream batches_file(absl::StrCat(data, "/batches_seed", seed, ".bin"),
                             std::ios::binary);
  if (!batches_file) open_spiel::SpielFatalError("No batches for this seed");
  std::vector<int32_t> indices;
  for (int32_t i; batches_file.read(reinterpret_cast<char*>(&i), sizeof(i));) {
    indices.push_back(i);  // Little-endian, as this machine.
  }
  const int steps = summary.at("steps").GetInt();
  SPIEL_CHECK_EQ(indices.size(), steps * batch_size);

  const std::string model_dir = absl::StrCat(data, "/model_seed", seed);
  std::filesystem::create_directories(model_dir);
  SPIEL_CHECK_TRUE(open_spiel::algorithms::torch_az::CreateGraphDef(
      *game, kLearningRate, kWeightDecay, model_dir, "vpnet.pb", "resnet", kNnWidth,
      kNnDepth));
  torch::manual_seed(seed);  // The weights are initialised in the constructor.
  VPNetModel model(*game, model_dir, "vpnet.pb", "/cpu:0");

  std::cout << absl::StrFormat(
                   "{\"variant\": \"cpp\", \"seed\": %d, \"train\": %d, \"test\": %d, "
                   "\"steps\": %d, \"checksums\": \"match the Python export\"}",
                   seed, train.size(), test.size(), steps)
            << std::endl;
  const auto start = std::chrono::steady_clock::now();
  for (int step = 1; step <= steps; ++step) {
    std::vector<VPNetModel::TrainInputs> inputs;
    for (int j = 0; j < batch_size; ++j) {
      inputs.push_back(Target(train[indices[(step - 1) * batch_size + j]]));
    }
    VPNetModel::LossInfo losses = model.Learn(inputs);
    if (step % kEvalEvery == 0) {
      const double seconds = std::chrono::duration<double>(
          std::chrono::steady_clock::now() - start).count();
      std::cout << absl::StrFormat(
                       "{\"variant\": \"cpp\", \"seed\": %d, \"step\": %d, "
                       "\"seconds\": %.0f, \"loss_policy\": %.4f, \"loss_value\": "
                       "%.4f, %s}",
                       seed, step, seconds, losses.Policy(), losses.Value(),
                       Evaluate(model, test))
                << std::endl;
    }
  }
  return 0;
}
